File: src/main/java/sh/miles/cosmictools/CosmicTools.java
package sh.miles.cosmictools;

import joptsimple.OptionSet;
import sh.miles.cosmictools.stage.CosmicReachDecompileStage;
import sh.miles.cosmictools.stage.CosmicReachDownloadStage;
import sh.miles.cosmictools.stage.CosmicReachUnzipStage;
import sh.miles.cosmictools.stage.MavenDownloadStage;
import sh.miles.cosmictools.stage.MavenInstallStage;
import sh.miles.cosmictools.stage.RunStage;
import sh.miles.cosmictools.stage.VineFlowerDownloadStage;
import sh.miles.cosmictools.util.NeoConstants;

import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/**
 * The main class for CosmicTools
 */
public final class CosmicTools {

    private static final LinkedList<RunStage> stages = new LinkedList<>();

    static {
        stages.add(new MavenDownloadStage());
        stages.add(new CosmicReachDownloadStage());
        stages.add(new CosmicReachUnzipStage());
        stages.add(new VineFlowerDownloadStage());
        stages.add(new CosmicReachDecompileStage());
        stages.add(new MavenInstallStage());
    }

    /**
     * The entry point of CosmicTools
     *
     * @param args the main arguments
     */
    public static void main(String[] args) {
        final long start = System.currentTimeMillis();
        final OptionSet options = NeoFlags.PARSER.parse(args);

        if (options.has(NeoFlags.HELP)) {
            try {
                NeoFlags.PARSER.printHelpOn(System.out);
            } catch (IOException ignore) {
            }
            System.exit(NeoConstants.SUCCESS);
        }

        final Map<String, Object> propagationMap = new HashMap<>();
        int exitCode = NeoConstants.SUCCESS;
        try {
            for (final RunStage stage : stages) {
                exitCode = stage.runStage(options, propagationMap);
                if (exitCode == NeoConstants.FAILURE) {
                    break;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(NeoConstants.FAILURE);
        }
        final long end = System.currentTimeMillis();
        System.out.printf("Finished Running CosmicTools in %d seconds%n", (end - start) / 1000);

        System.exit(exitCode);
    }

}

File: src/main/kotlin/sh/miles/cosmictools/CosmicTools.kt
package sh.miles.cosmictools

import sh.miles.cosmictools.runstage.CosmicReachDecompileRunStage
import sh.miles.cosmictools.runstage.CosmicReachDownloadRunStage
import sh.miles.cosmictools.runstage.DeleteCachesRunStage
import sh.miles.cosmictools.runstage.MavenDownloadRunStage
import sh.miles.cosmictools.runstage.MavenInstallRunStage
import sh.miles.cosmictools.runstage.RunStage
import sh.miles.cosmictools.utils.CurrentDirectory
import java.nio.file.Path
import kotlin.system.exitProcess

fun main(args: Array<String>) {
    val options = PARSER.parse(*args)

    if (options.has(HELP)) andSucceed { PARSER.printHelpOn(System.out) }
    if (options.has(CURRENT_WORKING_DIRECTORY)) {
        CurrentDirectory.cwd(Path.of(CURRENT_WORKING_DIRECTORY.value(options)))
    }

    val stages = setupStages()
    val propagate = mutableMapOf<String, Any>()
    for (stage in stages) {
        val exitCode = stage.runStage(options, propagate)
        if (exitCode == FAILURE) {
            exitProcess(FAILURE)
        } else if (exitCode == EXIT_NOW) {
            exitProcess(EXIT_NOW)
        }
    }
}

private fun setupStages(): List<RunStage> {
    return listOf(
        DeleteCachesRunStage(),
        CosmicReachDownloadRunStage(),
        CosmicReachDecompileRunStage(),
        MavenDownloadRunStage(),
        MavenInstallRunStage()
    )
}
